src/Microsoft.AspNet.Authentication/DataHandler/TicketSerializer.cs

```diff
@@ -36,98 +39,177 @@ public virtual AuthenticationTicket Deserialize(byte[] data)
             {
-                var authenticationType = reader.ReadString();
-                var nameClaimType = ReadWithDefault(reader, DefaultValues.NameClaimType);
-                var roleClaimType = ReadWithDefault(reader, DefaultValues.RoleClaimType);
-                var count = reader.ReadInt32();
-                var claims = new Claim[count];
-                for (int index = 0; index != count; ++index)
-                {
-                    var type = ReadWithDefault(reader, nameClaimType);
-                    var value = reader.ReadString();
-                    var valueType = ReadWithDefault(reader, DefaultValues.StringValueType);
-                    var issuer = ReadWithDefault(reader, DefaultValues.LocalAuthority);
-                    var originalIssuer = ReadWithDefault(reader, issuer);
-                    claims[index] = new Claim(type, value, valueType, issuer, originalIssuer);
-                }
-                var identity = new ClaimsIdentity(claims, authenticationType, nameClaimType, roleClaimType);
-                var bootstrapContextSize = reader.ReadInt32();
-                if (bootstrapContextSize > 0)
-                {
-                    identity.BootstrapContext = reader.ReadString();
-                }
-                identities[i] = identity;
+                identities[index] = ReadIdentity(reader);
+            }
+
+            var properties = PropertiesSerializer.Default.Read(reader);
+
+            return new AuthenticationTicket(new ClaimsPrincipal(identities), properties, scheme);
+        }
+
+        public virtual ClaimsIdentity ReadIdentity([NotNull] BinaryReader reader)
+        {
+            var authenticationType = reader.ReadString();
+            var nameClaimType = ReadWithDefault(reader, ClaimsIdentity.DefaultNameClaimType);
+            var roleClaimType = ReadWithDefault(reader, ClaimsIdentity.DefaultRoleClaimType);
+
+            // Read the number of claims contained
+            // in the serialized identity.
+            var count = reader.ReadInt32();
+
+            var identity = new ClaimsIdentity(authenticationType, nameClaimType, roleClaimType);
+
+            for (int index = 0; index != count; ++index)
+            {
+                var claim = ReadClaim(reader, identity);
+
+                identity.AddClaim(claim);
+            }
+
+            // Determine whether the identity
+            // has a bootstrap context attached.
+            if (reader.ReadBoolean())
+            {
+                identity.BootstrapContext = reader.ReadString();
             }
 
-            var properties = PropertiesSerializer.Read(reader);
-            return new AuthenticationTicket(new ClaimsPrincipal(identities), properties, authenticationScheme);
+            // Determine whether the identity
+            // has an actor identity attached.
+            if (reader.ReadBoolean())
+            {
+                identity.Actor = ReadIdentity(reader);
```

Comment:
Note: `ClaimsIdentity.Actor` is guarded against circular references, so there's no risk to end up in an endless recursive call :smile: